src/plugins/github/graphql.js

```diff
@@ -46,9 +119,235 @@ export function createQuery(): Body {
+ * choices as to where the boundaries should be. (For instance, should
+ * we include the type of the node that we want to fetch more of, or the
+ * type of the enclosing node? What sort of path information should we
+ * retain?)
+ */
+export function continuationsFromContinuation(
+  result: any,
+  source: Continuation
+): Iterator<Continuation> {
+  const continuationsFromEnclosingType = {
+    REPOSITORY: continuationsFromRepository,
+    ISSUE: continuationsFromIssue,
+    PULL_REQUEST: continuationsFromPR,
+    PULL_REQUEST_REVIEW: continuationsFromReview,
+  }[source.enclosingNodeType];
+  return continuationsFromEnclosingType(
+    result,
+    source.enclosingNodeId,
+    source.destinationPath
+  );
+}
+
+function* continuationsFromRepository(
+  result: any,
+  nodeId: string,
+  path: $ReadOnlyArray<string | number>
+): Iterator<Continuation> {
+  const b = build;
+  if (result.issues && result.issues.pageInfo.hasNextPage) {
+    yield {
+      enclosingNodeType: "REPOSITORY",
+      enclosingNodeId: nodeId,
+      selections: [
+        b.inlineFragment("Repository", [
+          b.field(
+            "issues",
+            {
+              first: b.literal(PAGE_LIMIT),
+              after: b.literal(result.issues.pageInfo.endCursor),
+            },
+            [b.fragmentSpread("issues")]
+          ),
+        ]),
+      ],
+      destinationPath: path,
+    };
+  }
+  if (result.pullRequests && result.pullRequests.pageInfo.hasNextPage) {
+    yield {
+      enclosingNodeType: "REPOSITORY",
+      enclosingNodeId: nodeId,
+      selections: [
+        b.inlineFragment("Repository", [
+          b.field(
+            "pullRequests",
+            {
+              first: b.literal(PAGE_LIMIT),
+              after: b.literal(result.pullRequests.pageInfo.endCursor),
+            },
+            [b.fragmentSpread("prs")]
+          ),
+        ]),
+      ],
+      destinationPath: path,
+    };
+  }
+  if (result.issues) {
+    for (let i = 0; i < result.issues.nodes.length; i++) {
+      const issue = result.issues.nodes[i];
+      const subpath = [...path, "issues", "nodes", i];
+      yield* continuationsFromIssue(issue, issue.id, subpath);
+    }
+  }
+  if (result.pullRequests) {
+    for (let i = 0; i < result.pullRequests.nodes.length; i++) {
+      const pr = result.pullRequests.nodes[i];
+      const subpath = [...path, "pullRequests", "nodes", i];
+      yield* continuationsFromPR(pr, pr.id, subpath);
+    }
+  }
+}
+
+function* continuationsFromIssue(
+  result: any,
+  nodeId: string,
+  path: $ReadOnlyArray<string | number>
+): Iterator<Continuation> {
+  const b = build;
+  if (result.comments.pageInfo.hasNextPage) {
+    yield {
+      enclosingNodeType: "ISSUE",
+      enclosingNodeId: nodeId,
+      selections: [
+        b.inlineFragment("Issue", [
+          b.field(
+            "comments",
+            {
+              first: b.literal(PAGE_LIMIT),
```

Comment:
I appreciate that once we hit a continuation, we use the full PAGE_LIMIT rather than the per-type page limit.

(Although, is it hypothetically possible that we we would over-use our total pages limit? what's the maximum number of query fragments we could be putting into a single continuation? is it 100^(max enclosing depth) if we got more-than-full pages on every level of the nesting?)

if we did hit this issue, would it fail clearly? i'm curious, but not too worried in practice.

Reply:
It’s not 100^(depth), because we only use the full `PAGE_LIMIT` on the outermost continuation. Within the continuation selections, we use the normal fragments, which have the shorter limits.

I think it is not too likely that we will over-use the total entities-per-page limit of 500K. Note that we can statically determine whether this might happen. If we do hit this issue, it will fail clearly: the GitHub API will return a clear message, we’ll catch it in the `errors` of the result, and reject in the `postQuery` handler in `fetchGithubQuery.js`.